M/N multisig support

naughtyfox opened the pull request:
Extended multisignatures to schemes M/N.

I saved compatibility with previous API flow for multisig wallets creation. 2/4 wallets creation process looks as follows:

```
[wallet 5BUgPa]: prepare_multisig
Wallet password: 
MultisigV1XQAM98AQLYTAR83FDuGbn3Crvtgm9jRyKKEryVwcHC5j8sCaryT339hNBrbWXQUaEe4NUB7vv9BdPHTuZuUFvSoVbC273qbGrJGcAcDHbP91oXXd46xUVSzknRDQQ8L9ddjmZX789VveWRYUeyrP6Q5UVpEzmDn1NpfRA4NA21UxihEx
Send this multisig info to all other participants, then use make_multisig <threshold> <info1> [<info2>...] with others' multisig info
This includes the PRIVATE view key, so needs to be disclosed only to that multisig wallet's participants 

[wallet 5BUgPa]: make_multisig 2 MultisigV1A9TsWdMbPHtgvM6FDuYR7nPovfNdTPteSPFZwFgcNeR6PNsrmGjPvHbhiq7Ra1sMs9VrW5ev5MBX5aP1zTXmFLPcRcnj8i4zvoJ7qZS7YqyGdSgXfvwcoqS3bbtmASNmHivZNCyJkno5mdB1JNxD5wqUtNVUiQwRRfifTK6WFaNZjE2H MultisigV1hzHY9B7TuKaMF6cjjA53j4418G164HFzuVjtKGiam5YaVJm2rgTnXWZSHsQgPwW4abXpTS74SB8KbNs69DfzrsWB69ghrcRz3xuDUi4nnwEvxmAk8ECQQicbXiVfuWPyijdGUuiRHvnGBw6G3nXiWTYoKE5oLLgo7iGryFi68Vrp9MT2 MultisigV1SKvVDZ1ReJXMDbugtaJBiEEQ9hCVGD51we3b7w4pV5bPTCoigR7G2qoMQcaAgp87pXZCY9wDsHbRUHGwJYaGQ1NofLJJVP7they2rM348cpjEgAPVUoLxa52wjMP7pZQsfMeBCpZWKtiPL91J7R7uWes2DYH1gBRgpsHhFVjEeGrEDzx
Wallet password: 
Another step is needed
MultisigxV1jag57V1msEsWDDtYtTmaBCAR97maah4H2hxLeXfrXRWUEtNigxghaLA1ZjmjYiTyjYSXTucNMEeNvLBKtqrUME5ZNvtX3iiLLmJCzoXAMn2YhYDTUEpVi1bMN8fv3oxE9tQWNbghwVYazvjgBhXcRoNuh989GT84GEuDpjPcfSYDrosPeGj8KKH1AmsbQPCNoA7ppi6pSwjrzxJJW436KbKqHBwkAkhQVevWjj8Gmsbgg5ZQZcFRfFpMbJp8DXEtX32rKzEP
Send this multisig info to all other participants, then use exchage_multisig_keys <info1> [<info2>...] with others' multisig info

[wallet 51sLpF (out of sync)]: exchange_multisig_keys MultisigxV16oHHLiL2WsyEsmRj1fc37YJTmQs4bgmERBGRTKSoFxtR253qC31hnnTE7wy8xLFumwWmQzyAdjJjzR2cwfatJkpU6HX6WFSqE6DDsn5xyrPNiXifA6QCSfZxycnGA1roQ76p79BWRqr7yFzWUKu2uZYMDDYre85YEyZp5emiVz2JXeakQwARLCMk9neSSUsZFbPmPcMNmby9gQvScPgTXmKeBUKAjBK7njF6M1hGEkcKdPT2caiSYaX8V1mrZGpSxaESWDV9 MultisigxV1P5bu44mS6SqLWfh5f8AuNzDotbMT6iEf7LVaFfixJnyuM2VBNiwcJSbBbFifbgj7VidFZjzV1kpi64gRQk1aigGS6HX6WFSqE6DDsn5xyrPNiXifA6QCSfZxycnGA1roQ76pGSNmJhNYMSUNTdB2HviM6hFL2E9Uypy5eMjuQ9B4Wx4oBpXWmfjGbTk1qBUqPhyWwwU58jMqa9oQLSbv9TX8T82CYdFUEEgVVAvM1hoHjzpofLLxzPKFjyCofZ2UK3HsoLeq MultisigxV1BjP86fURDjY2L31oxTC64GDfDdPfqogahNdYLw2evHbo8PYmCtWxM8u7wSb6LY2VVrR4zW7vpq7QkGGkV2mYh19M79BWRqr7yFzWUKu2uZYMDDYre85YEyZp5emiVz2JXeakGSNmJhNYMSUNTdB2HviM6hFL2E9Uypy5eMjuQ9B4Wx4oE3351oRkZD86TH4z5SjThn2do9u6UXgDRExMNatSRcKVUCKFnmLLdCwdLYJATYMc2wFgq7rVG2LMdMX8jffrgVpC
Wallet password: 
Another step is needed
MultisigxV1AExg753mNvX17iyTbFX6Tq3sjpJ2YcPsgWi9jQkM2rMKctP8R5sBzZeCRDPXmCvHF54Gs7BKCj7W6AgPiwaLEJVF1r8MKR6yUmkeJMfexLhwun6VGqErFWFMV2SY1wXK5jNvFqLpiK3Ng6QjiWnAiqDccLajgFZvT5uLrXFgJm8JgrF5domBJYwSgZ89sfeC2RP45uQq6ksxGL3pYBThLwoWNieJS1A7P4MRqiiCYc2NqstSfzRVBeEZ3ezGhGNF89E2zL2YJos52A14AdFdF6PNUdK7UhMuGntMeBH8bgKAAuz1bfa57JDe4dwoC1rRTuhvL2Ewhq49PaQeBtntqHJuDnHwYme6cGkJGrEJoMqaX9tSajc1FkBgBop7hvkC6cq8BtMCjmZT
Send this multisig info to all other participants, then use exchange_multisig_keys <info1> [<info2>...] with others' multisig info

[wallet 51sLpF]: exchange_multisig_keys MultisigxV1PUq5CDZ9oWf3jUcWdx8aC32pWeCySeE7wJJH1aFaPVyAetYD2Bm9yh4PUSNuLtScvV2NKZzbQWQCgJq5saoLvCmVh5h5WtcM25Pc23k5ovZFjY8LnRjfT4DbkTCHv7kh2oPPRsW3Ax9XKFFEjRPh5DECjzCQMWwKXa9KneZmiQWB6kohV9nAHzG3ohYcHQq8E9NZR8UxsMDjRpmTxj6FrDPFCj2DXavpoe9reN94owQvP1dCc1XacHsVg7RKSKQMgUj8DNnS MultisigxV1dLf4xkCR7tD7C2cRhR1YATW5yJVFm2XiANVuMQjwsiM2etYD2Bm9yh4PUSNuLtScvV2NKZzbQWQCgJq5saoLvCmVd6syGgSUGA5hh8tkn8A6vpNhbwP5PjfBhfREypkSWuXjRsW3Ax9XKFFEjRPh5DECjzCQMWwKXa9KneZmiQWB6kohRuXSZGmcDNxYQoS5WSvC3P2qKQLTvLxULCDXUqbz6rJzA1st1tg4iRVdjdiWbyVPKRSBSU9oCx3wKKUkqhjqCGJR MultisigxV16FJqCGcFCnBTNTQmbY3CN7C5HqW62EMhXAh6Nax6rZ6UetYD2Bm9yh4PUSNuLtScvV2NKZzbQWQCgJq5saoLvCmVh5h5WtcM25Pc23k5ovZFjY8LnRjfT4DbkTCHv7kh2oPPd6syGgSUGA5hh8tkn8A6vpNhbwP5PjfBhfREypkSWuXj6tQM6LDnKbnJ8QkB9CePUK7q7wYM7x8JbJmj4D26Mko2DYAB5vbtmYjZxeapFTHuWYNgcgB4KAqoLFTb4M5Z9eGo
Wallet password: 
Multisig wallet has been successfully created. Current wallet type: 2/4
[wallet 526g4j]: address
0  526g4jUk2MxYjKFGrq655vAC9CBYLM66sWLn1VzXeAukE52fUbUkjsecpLLAzgBt7C9mTkutf29fnUEpPt39KUYZJ5votVx  Primary address 
```

For classic `N-1/N` wallets instead of `exchange_multisig_keys` you _may_ also use `finalize_multisig` command as you did before. Signing and submitting multisignature transactions procedure as well as partial key images exchange remain the same. In other words, only `exchange_multisig_keys` was added from API perspective.

Support for arbitrary scheme has been built also into:
- wallet api
- waller rpc
- `monero-gen-trusted-multisig` utility

Two members `m_multisig_rounds_passed` and `m_multisig_derivations` have been added in `wallet2` class to be able to continue creation procedure if the wallet has been closed between rounds.

On of the first transactions sent from 2/4 wallet on _stagenet_ is `244697a55e460e2188bb43b7a95ecf51298de1b6bc6a81f88d6d170e943ad0ef`, from 2/5 wallet - `fd3bb0a12e44cdc4aa7834032c9ac15c7b0528c0bf9644d851e4d99d47606368`. I may provide these wallets if needed :)

Tested with wallet api, wallet cli and wallet rpc on Ubuntu 17.10.
Core tests have also been added.

naughtyfox commented:
@moneromooo-monero @stoffu and @luigi1111 please review my changes

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3466,72 +3536,69 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+      extra_multisig_info = pack_multisignature_keys(MULTISIG_EXTRA_INFO_MAGIC, derivations, m_account.get_keys().m_spend_secret_key);
+      spend_skey = rct::sk2rct(m_account.get_keys().m_spend_secret_key);
+      m_multisig_derivations = derivations;
+    }
```
Why is N-1/N different from N-x/N with x>1 here ?

naughtyfox replied:
Because of the hashing of keys derivation on last round. It breaks the property of DH: `B0 * b1 * b2 = b0 * B1 * b2 = b0 * b1 * B2`. We talked about and I asked @luigi1111 about it. As far as I remember he agreed.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3466,72 +3536,69 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
   MINFO("Creating multisig address...");
   CHECK_AND_ASSERT_THROW_MES(m_account.make_multisig(view_skey, rct::rct2sk(spend_skey), rct::rct2pk(spend_pkey), multisig_keys),
-      "Failed to create multisig wallet due to bad keys");
+    "Failed to create multisig wallet due to bad keys");
```
spurious

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3557,17 +3624,140 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+
+  THROW_WALLET_EXCEPTION_IF(false,
+    error::wallet_internal_error, "Unsupported info string");
+  return {};
```
Looks a lot simpler to do that in the test above

naughtyfox replied:
You totally right here :)

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3557,17 +3624,140 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+    // the last round is passed and we have to calculate spend public key
+    // add ours if not included
+    crypto::public_key local_signer;
+    CHECK_AND_ASSERT_THROW_MES(crypto::secret_key_to_public_key(get_account().get_keys().m_spend_secret_key, local_signer),
```
There is a function for that.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3557,17 +3624,140 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+
+    m_subaddresses.clear();
+    m_subaddress_labels.clear();
+    add_subaddress_account(tr("Primary account"));
```
Init m_last_block_reward too (new thing, check other generate calls)

naughtyfox replied:
It's already inited in `make_multisig`

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3557,17 +3624,140 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+    MINFO("Preparing keys for next exchange round...");
+    extra_multisig_info = pack_multisignature_keys(MULTISIG_EXTRA_INFO_MAGIC, new_derivations, m_account.get_keys().m_spend_secret_key);
+    m_multisig_derivations = new_derivations;
+  }
```
Why are those two cases different ?

naughtyfox replied:
All additional rounds that happen in N-x/N are just DH key exchange. Last round is for generating common public spend key only. And before performing last round we generate our full private secret key and deriving multisig public keys to calculate full public spend key. We don't need to do it in intermediate rounds. So that's the difference.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3557,17 +3624,140 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
   {
     THROW_WALLET_EXCEPTION_IF(!verify_multisig_info(info[i], secret_keys[i], public_keys[i]),
-        error::wallet_internal_error, "Bad multisig info: " + info[i]);
+      error::wallet_internal_error, "Bad multisig info: " + info[i]);
```
spurious

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3604,74 +3794,54 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
     {
       THROW_WALLET_EXCEPTION_IF(public_keys[i] == local_pkey, error::wallet_internal_error,
-          "Found local spend public key, but not local view secret key - something very weird");
+        "Found local spend public key, but not local view secret key - something very weird");
```
Is your editor munging these automatically ? Must be a setting to stop it.

naughtyfox replied:
Yes, it's editor's fault. The project uses different identations and my IDE goes crazy about it

moneromooo-monero commented on src/wallet/wallet_rpc_server_commands_defs.h:
```diff
@@ -1916,6 +1916,31 @@ namespace wallet_rpc
     };
   };
 
+  struct COMMAND_RPC_EXCHANGE_MULTISIG_KEYS
```
Bump RPC version at the top

naughtyfox replied:
done

moneromooo-monero commented on tests/core_tests/chaingen.h:
```diff
@@ -549,7 +549,7 @@ inline bool do_replay_file(const std::string& filename)
         for (const auto &k: all_multisig_keys) \
           spend_public_keys.push_back(k); \
-        crypto::public_key spend_pkey = cryptonote::generate_multisig_N1_N_spend_public_key(spend_public_keys); \
+        crypto::public_key spend_pkey = cryptonote::generate_multisig_M_N_spend_public_key(spend_public_keys); \
```
Can you add some tests for the new threshold cases ?

naughtyfox replied:
I can... But the problem is core tests don't use wallet or it's parts and I need to copy and paste some code into appropriate macro...

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -3557,17 +3624,140 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+  CHECK_AND_ASSERT_THROW_MES(!derivations.empty(), "empty pkeys");
+  bool ready = false;
+  CHECK_AND_ASSERT_THROW_MES(multisig(&ready), "The wallet is not multisig");
+  CHECK_AND_ASSERT_THROW_MES(!ready, "Multisig wallet creation process has already been finished");
```
Probably should be a sanity check for signers size here.

naughtyfox commented on src/multisig/multisig.cpp:
```diff
@@ -81,6 +81,43 @@ namespace cryptonote
     }
   }
   //-----------------------------------------------------------------
+  std::vector<crypto::public_key> generate_multisig_derivations(const account_keys &keys, const std::vector<crypto::public_key> &derivations)
```
because `sk * Pk = Pk`

moneromooo-monero replied:
I do not understand the point.

moneromooo-monero replied:
It does not generate derivations though, but public keys. Maybe you call these public keys derivations in your scheme ? If so, it's just confusing :)

naughtyfox replied:
I always thought about key derivations as functions spawning keys from another keys. In our case spawned key is public. Maybe I was wrong :) Do you want my to rename it?

naughtyfox commented on src/gen_multisig/gen_multisig.cpp:
```diff
@@ -143,11 +143,7 @@ static bool generate_multisig(uint32_t threshold, uint32_t total, const std::str
-          tools::fail_msg_writer() << genms::tr("Error finalizing multisig");
-          return false;
-        }
+          extra_info[n] = wallets[n]->exchange_multisig_keys(pwd_container->password(), pkeys, signers);
```
because it throws in case of error

naughtyfox commented on src/multisig/multisig.cpp:
```diff
@@ -138,4 +175,8 @@ namespace cryptonote
   }
   //-----------------------------------------------------------------
+  uint32_t multisig_rounds_required(uint32_t participants, uint32_t threshold)
+  {
```
this case is checked before this function called. but i can just in case

naughtyfox replied:
done

naughtyfox commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -958,6 +958,61 @@ bool simple_wallet::finalize_multisig(const std::vector<std::string> &args)
+        return true;
+      } else {
+        uint32_t threshold, total;
+        m_wallet->multisig(NULL, &threshold, &total);
```
I did the check at the first then decided it's redundant because we already have check the wallet is multisig above. Do you think we need to check it here again?

naughtyfox commented on src/wallet/wallet2.h:
```diff
@@ -558,6 +558,18 @@ namespace tools
+     * \param password
+     * \param pkeys
+     * \param signers
+     * \return
```
sorry, it's autogenerated :)

naughtyfox commented on src/wallet/wallet2.cpp:
```diff
@@ -2938,6 +2990,24 @@ bool wallet2::load_keys(const std::string& keys_file_name, const epee::wipeable_
+          return false;
+        }
+        const char *field_multisig_derivations = json["multisig_derivations"].GetString();
+        std::string multisig_derivations = std::string(field_multisig_derivations, field_multisig_derivations + json["multisig_derivations"].GetStringLength());
```
I wasn't sure, I just followed the pattern used couple of strings above

naughtyfox commented on src/wallet/wallet2.cpp:
```diff
@@ -133,6 +134,42 @@ namespace
+  {
+    std::string data;
+    crypto::public_key signer;
+    CHECK_AND_ASSERT_THROW_MES(crypto::secret_key_to_public_key(signer_secret_key, signer), "Failed to derive public spend key");
```
`pack_multisignature_keys(const std::string& prefix, const std::vector<crypto::public_key>& keys, const crypto::secret_key& signer_secret_key)` function used before we assigned spend secret key, I don't think it would be relevant here...

naughtyfox commented on src/wallet/wallet2.cpp:
```diff
@@ -3466,72 +3536,69 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
+    multisig_signers = spend_keys;
+
+    crypto::public_key signer;
+    CHECK_AND_ASSERT_THROW_MES(crypto::secret_key_to_public_key(get_account().get_keys().m_spend_secret_key, signer), "Failed to generate signer public key");
```
using this function here would throw because at the moment is not in multisig state

moneromooo-monero replied:
That sounds either unlikely, or a hint of a bug somewhere.

naughtyfox replied:
Why does it sound unlikely? `get_multisig_signer_public_key` meant to be used when wallet creation is already finished and in `multisig` state.

At the moment we can calculate secret spend key for each participant only in case of classic N/N or N-1/N schemes and get permanent signer public key. In other cases secret spend can be calculated  only before last round.

If you want this key to be derived with `get_multisig_signer_public_key` I can remove the check inside of it. Btw, in previous version of code this function hasn't been used either. 

I can't see buggy behavior here for now. If you see better solution please let me know

naughtyfox replied:
okay, will be done

naughtyfox commented on src/wallet/wallet2.cpp:
```diff
@@ -3466,72 +3536,69 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
   std::string extra_multisig_info;
-  crypto::hash hash;
-
-  clear();
```
I moved it because calling `clear()` at that moment wasn't exception safe - in case of error during calculations whole wallet would get in invalid state

naughtyfox commented:
I fixed part of comments, parts of comments need to be discussed. I will add some core tests soon.

naughtyfox commented:
@moneromooo-monero I added core tests for my functionality as promised. Please check it out. I also fixed some of comments, others need to be discussed further :)

naughtyfox commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -887,7 +887,7 @@ bool simple_wallet::make_multisig(const std::vector<std::string> &args)
       success_msg_writer() << tr("Another step is needed");
       success_msg_writer() << multisig_extra_info;
-      success_msg_writer() << tr("Send this multisig info to all other participants, then use finalize_multisig <info1> [<info2>...] with others' multisig info");
+      success_msg_writer() << tr("Send this multisig info to all other participants, then use exchage_multisig_keys <info1> [<info2>...] with others' multisig info");
```
fixed, but github doesn't show it some reason

ph4r05 commented:
@naughtyfox I was trying to find a (white)paper describing the M/N multisig scheme you implemented but I haven't found it. Can you pls link the paper with the scheme? 

Thanks!

naughtyfox commented:
@ph4r05 unfortunately I couldn't find public multisig paper either. We discussed details about extending it to arbitrary M/N scheme in `#monero-research-lab` IRC channel. The only part of my work here is to extend key exchange rounds to support M/N. This is enough to be integrated in partial key images exchange and MLSTAG signatures. I can do a write up to describe better what happens under the hood with wallet's keys. Would it be ok for you?

ph4r05 commented:
@naughtyfox Aaah I see. 

I think the write-up describing how exactly the multisig variants work (also with the M/N) would be very beneficial to the Monero community. Without following the IRC channels or "reverse-engineering" it from the source code it may be quite complicated to get an idea how multisig in Monero works.

IMO, Having published documentation would help to increase the transparency and security of the Monero, so other security researches could assess the multisig scheme security (and to check whether implementation matches the desired scheme).

Unfortunately, there is no MRL-xx document yet on the multisig topic, as far as I know.
There is a description in the @moneromooo-monero commit message introducing the Multisigs https://github.com/moneromooo-monero/bitmonero/commit/b99bd7eecf 

There is also a paper on eprint describing basic Monero transactions, but again without multisigs.
https://eprint.iacr.org/2018/535

naughtyfox commented:
This is true. 
Recently we tried to fill up the gap and issued post on hackernoon about basics in monero multisignatures - https://hackernoon.com/monero-multisignatures-explained-46b247b098a7. And of course, it's about N/N and N-1/N schemes only. 
I'll write some doc about multisignatures and will try to cover keys exchange and partial key images topics.
**PS** I almost forgot! @ukoehb wrote the book about monero internals and there is unpublished chapter about multisignatures - https://github.com/UkoeHB/Monero-RCT-report/blob/master/multisig_chapter-1-0.pdf

moneromooo-monero commented:
Surae's almost done with the multisig paper.
You can ask him for a link to the current draft.

b-g-goodell commented:
View access for current draft: https://v2.overleaf.com/read/bfjfkdgnhgvh we'll be submitting it for publication soon so I'm not tremendously willing to include the more general case *for this document*.. but I can be persuaded, especially if someone has a clearish write-up to include! I can't look at it very soon but I agree that at least the more general M/N description should at least end up in the Zero to Monero document

naughtyfox commented:
I made some kind of writing to explain in more human readable manner what's been done in this PR. Unfortunately, I don't have math-writing knack so I did it in google docs (again) in more or less free manner. You may check this out here - https://docs.google.com/document/d/1lo2oLyXP8O8g98FkNK9feUvZx98H6h9fgvBZn5-Qiwg/edit?usp=sharing

moneromooo-monero commented:
Please rebase.

b-g-goodell commented:
Multisig paper submitted to IACR. Link incoming in a few days.

naughtyfox commented:
rebased

naughtyfox commented:
Me and @b-g-goodell had a conversation about hashing multisig in the middle of key exchange rounds and why rounds differs as pointed in this discussion - https://github.com/monero-project/monero/pull/4036#discussion_r202344950. I wrote the doc to clarify why it's needed - https://docs.google.com/document/d/1CGzdx8AeaDgWI3HZ1HpYQDtsPY1iirqGfOjvrMhlpss/edit?usp=sharing

naughtyfox commented:
rebased

naughtyfox commented on src/wallet/wallet2.cpp:
```diff
@@ -3847,43 +3917,78 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
-  MINFO("Creating spend key...");
-  std::vector<crypto::secret_key> multisig_keys;
-  rct::key spend_pkey, spend_skey;
+  // In common multisig scheme there are 4 types of key exchange rounds:
```
Tried to explain how key exchange process goes

naughtyfox commented on src/wallet/api/wallet.cpp:
```diff
@@ -1187,6 +1187,20 @@ string WalletImpl::makeMultisig(const vector<string>& info, uint32_t threshold)
+        clearStatus();
+        checkMultisigWalletNotReady(m_wallet);
+
+        return m_wallet->exchange_multisig_keys(epee::wipeable_string(m_password), info);
```
Do I need to `SCOPED_WALLET_UNLOCK();` here as `simplewallet` does?

naughtyfox commented:
@moneromooo-monero @b-g-goodell I added some descriptive comments. Also added multisig unit tests for 2/4 and 2/5 schemes which is passed. Rewrote multisig wallets creation function in unit tests also to make it common and more simple (i'm ready to discuss it).

I also have some questions about when I should use keys unlockers like `SCOPED_WALLET_UNLOCK()` in wallet api calls? It's kinda not transparent for me.

moneromooo-monero commented:
You use SCOPED_WALLET_UNLOCK when you want the wallet keys to be unlocked till the end of the scope. You use LOCK_IDLE_SCOPE when you want to be the only one accessing wallet2 till the end of the scope. The former implies the latter.
